# Augur v6 release notes

Augur v6 was released on _TODO_.
This release contains a number of changes from augur v5, including feature additions & bugfixes.
The biggest change is related to how augur _exports_ files for visualisation by [auspice](https://nextstrain.github.io/auspice/).
We've written [an extensive guide](migrating-v5-v6) explaining our motivations here, whats changed, how to upgrade, and how this interfaces with auspice.

---

* [Export JSONs for specific versions of Auspice](#export-jsons-for-specific-versions-of-auspice)
* [Reference sequence output](#reference-sequence-output)
* [Change in `augur ancestral`s arguments](#change-in-augur-ancestrals-arguments)
* [Import BEAST MCC trees](#import-beast-mcc-trees)
* [Prettifying of strings](#prettifying-of-strings)
* [Whitespace in colors & lat-longs TSVs](#whitespace-in-colors-lat-longs-tsvs)
* [Move to GFF-style annotations](#move-to-gff-style-annotations)
* [Improvements & usage of JSON validation](#improvements-usage-of-json-validation)
* [Removal of non-modular augur & old builds](#removal-of-non-modular-augur-old-builds)
* [Test builds](#test-builds)
* [Documentation improvements](#documentation-improvements)
* [Miscellaneous](#miscellaneous)

---

## Export JSONs for specific versions of Auspice
Probably the biggest (breaking) change you'll encounter is that `augur export` no longer works!
See the [migration guide](migrating-v5-v6) for a detailed explanation of this.

<span style='color: orange'>Breaking change:</span> `augur export` no longer works, and now requires a further positional argument to define which version of auspice you wish to target.
`augur export v1` should behave the same as previous versions' `augur export`.



## Reference sequence output


## Change in `augur ancestral`s arguments

`augur ancestral`, which reconstructs mutations across a tree, now supports two forms of output, and the arguments have become more descriptive.

1. JSON output, including mutations for each branch & (inferred) ancestral sequences. 
This is specified by the `--output-node-data` argument.

2. FASTA output of reconstructed ancestral sequences.
This had previously been available for VCF-inputs, but now works for any input. 
Users can ask for this output and specify a file name using `--output-sequences`.

<span style='color: orange'>Deprecation warning:</span> The argument `--output` is now deprecated. Please use `--output-node-data` instead.

## Import BEAST MCC trees

## Prettifying of strings



## Whitespace in colors & lat-longs TSVs

## Move to GFF-style annotations

Starting with augur v6 we use GFF coordinates: [one-origin, inclusive] as opposed to
And strainds are represented by `+` or `-` rahter than `1` or `0`
And we export the `seqid`, but don't use it in auspice.

## Improvements & usage of JSON validation

## Removal of non-modular augur & old builds
Augur has been a dynamic, shapeshifting beast.
It started as scripts for nextflu, took on more & more pathogens, was refactored into "prepare" and "process" steps, and refactored again into the "modular" augur we now have.
Earlier incarnations of augur have now been removed from the GitHub repo (`./base/*`).

Parallaling the different incarnations of augur was a move to "builds" being their own self contained repos.
We think this has been remarkable successful, and decouples the bioinformatics tooling from a pathogen build.
With this release of augur we've now removed these builds from the augur GitHub repo, and the only builds that remain are the test ones...

## Test builds 

There have been a number of test builds in the augur repo and we have leaned heavily on them while we developed this version of augur as well as auspice v2.
They are all self contained within `./tests/builds` and can all be run & examined in auspice via

```bash
cd tests/builds
bash runner.sh # creates output in ./auspice
auspice view --datasetDir auspice
```

(See the [auspice docs](https://nextstrain.github.io/auspice) for auspice-specific questions.)

## Documentation improvements

Documentation has always been a bit hit-or-miss with nextstrain projects.
We've tried to make these (augur's read-the-docs documentation) more comprehensive and with better flow.
This entails new sections, with each augur command having its own page.
We've tried to use redirects to ensure that all the old links continue to work.

## Miscellaneous

* `augur filter`: More interpretable output of how many sequences each stage has filtered out.

* `augur sequence-traits`: Numerical output as originally intended, but required an auspice bugfix.

* `augur traits`: Explanation of what is considered missing data & how it is interpreted

* `augur traits`: GTR models are exported in the output JSON for better accountability & reproducibility.

* Errors in formatting of input files (e.g. metadata files, auspice config files) weren't handled nicely, often resulting in hard-to-interpret stack traces.
We now try to catch these and print an error indicating the offending file.

* Tests using python version 2 have now been removed.